### PR TITLE
fix: add an option to stopAll in compound launch configs

### LINKED ISSUE #95964: Debug fails to disconnect from targets
Ever since I started using the new JS debugger, when running our top level `VS Code` target and then manually closing the code-oss window, the `Attach to Shared Process` and `Attach to Extension Host` targets remain:

image

Eventually, maybe 1 minute later, the `Attach to Extension Host` goes away, but the other one stays, I always have to manually disconnect from it.

image

cc @weinand @isidorn

### PR DESCRIPTION
**Pull request by connor4312**
This PR fixes #95964

It adds a "stopAll" attribute which tears down other sessions if a single debug session is _manually_ terminated. If a session ends automatically/from the DA message, we won't tear it down.

### CODE REVIEW

File: src/vs/workbench/contrib/debug/common/debugCompoundRoot.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	onShouldSessionsStop = this.stopEmitter.event;
+
+	didStop(): void {
+		if (!this.stopped) { // avoid sending extranous terminate events
+			this.stopped = true;
+			this.stopEmitter.fire();
+		}
+	}
+}
+
```
Comment (isidorn) [inline src/vs/workbench/contrib/debug/common/debugCompoundRoot.ts:22]: stubCompoundRoot is nowwhere used.
We should remove it and when needed we should put it in `mockDebug.ts`

File: src/vs/workbench/contrib/debug/common/debugSchemas.ts [hunk 2]
```diff
@@ -215,6 +215,11 @@ export const launchSchema: IJSONSchema = {
 						},
 						description: nls.localize('app.launch.json.compounds.configurations', "Names of configurations that will be started as part of this compound.")
 					},
+					stopAll: {
+						type: 'boolean',
+						default: false,
+						description: nls.localize('app.launch.json.compound.stopAll', "Whether to stop all of the compound sessions when any session is terminated.")
```
Comment (isidorn) [inline src/vs/workbench/contrib/debug/common/debugSchemas.ts:221]: I suggest to use the word "controls" and "manual". Thus
"Controls if manually terminating one session will stop all of the compound sessions"
Though your native english is better, so just take this as a suggestion.

File: src/vs/workbench/contrib/debug/common/debugCompoundRoot.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { IDebugCompoundRoot } from 'vs/workbench/contrib/debug/common/debug';
+import { Emitter } from 'vs/base/common/event';
+
+export class DebugCompoundRoot implements IDebugCompoundRoot {
+	private stopped = false;
+	private stopEmitter = new Emitter<void>();
+
```
Comment (isidorn) [inline src/vs/workbench/contrib/debug/common/debugCompoundRoot.ts:12]: We usualy name our events like what happend, not what should happen as a consequence.
Thus I suggest the following name `onDidEndContainingSession` or `onDidEndChildSession` or just `onDidEndSession`. To be algined with the existing `onDidEndSession` that lives in the `debugService`

File: src/vs/workbench/contrib/debug/common/debugCompoundRoot.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+	onShouldSessionsStop = this.stopEmitter.event;
+
+	didStop(): void {
```
Comment (isidorn) [inline src/vs/workbench/contrib/debug/common/debugCompoundRoot.ts:15]: `didStop` sounds like a question? I suggest to call it `sessionStopped` or something like that.

File: src/vs/workbench/contrib/debug/common/debug.ts [hunk 5]
```diff
@@ -910,3 +912,9 @@ export interface IDebugHelperService {
 
 	createTelemetryService(configurationService: IConfigurationService, args: string[]): TelemetryService | undefined;
 }
+
+export interface IDebugCompoundRoot {
```
Comment (isidorn) [inline src/vs/workbench/contrib/debug/common/debug.ts:916]: I do not see the benefits of introducing this interface.
Though we can have it if you prefer it that way.

**Review by isidorn — CHANGES_REQUESTED** [review]
Great work. Added some comments, mostly around naming.

File: src/vs/workbench/contrib/debug/common/debug.ts (hunk 5 above)
Comment (connor4312) [reply src/vs/workbench/contrib/debug/common/debug.ts:916]: Yea I originally had the compound root as required and used the stub, but we no longer do that so I'll clean it up!

**Review by isidorn — APPROVED** [review]
Feel free to merge! Thanks a lot for this work!

**Pull request merged**